File: python/BUILDER_COMPOSITE/BUILDER_COMPOSITE/builder/ConstructorReservacionMultiple.py
from ..Reservacion.ReservacionSencilla import ReservacionSencilla
from ..Reservacion.ReservacionVIP import ReservacionVIP
from ..composite.ReservacionMultiple import ReservacionMultiple
from .ConstructorReservacion import ConstructorReservacion

class ConstructorReservacionMultiple(ConstructorReservacion):
    
    def __init__(self) -> None:
        self.reservacionMultiple = ReservacionMultiple()
    
    def construirDatosCliente(self, nombres,
                              apellidoPaterno,
                              apellidoMaterno,
                              edad,
                              tipoCliente):
        self.reservacionMultiple.datosCliente.setDatosCliente(nombres,
                              apellidoPaterno,
                              apellidoMaterno,
                              edad,
                              tipoCliente)
    
    def construirDatosCompra(self, numeroTarjeta,
                             fechaVencimiento):
        self.reservacionMultiple.datosCompra.setDatosCompra(numeroTarjeta,
                             fechaVencimiento)
    
    
    def construirDatosFuncion(self, asiento,
                              horario,
                              pelicula):
        self.reservacionMultiple.datosFuncion.setDatosFuncion(asiento,
                              horario,
                              pelicula)
    
    def construirDatosBeneficios(self, beneficios):
        if beneficios is not None:
            self.reservacionMultiple.datosBeneficios.setDatosBeneficios(beneficios)      
            reservacion = ReservacionVIP()
            
            reservacion.datosCliente = self.reservacionMultiple.datosCliente
            reservacion.datosCompra = self.reservacionMultiple.datosCompra
            reservacion.datosFuncion = self.reservacionMultiple.datosFuncion
            reservacion.datosBeneficios = self.reservacionMultiple.datosBeneficios
            
            self.reservacionMultiple.reservaciones.append(reservacion)
        else: 
            reservacion = ReservacionSencilla()
            reservacion.datosCliente = self.reservacionMultiple.datosCliente
            reservacion.datosCompra = self.reservacionMultiple.datosCompra
            reservacion.datosFuncion = self.reservacionMultiple.datosFuncion
            
            self.reservacionMultiple.reservaciones.append(reservacion)
    
    def obtenerReservacion(self):
        return self.reservacionMultiple
        

File: php/BUILDER_COMPOSITE/constructores/ConstructorReservacionMultiple.php
<?php

require_once './constructores/ConstructorReservacion.php';
require_once './reservaciones/ReservacionSencilla.php';
require_once './reservaciones/ReservacionVIP.php';
require_once './reservaciones/ReservacionMultiple.php';


class ConstructorReservacionMultiple implements ConstructorReservacion{
    
    public $reservacionMultiple;
    
    public function __construct() {
        $this->reservacionMultiple = new ReservacionMultiple();
    }
    
    public function construirDatosCliente($nombre, $apellidoP, $apellidoM, $edad, $tipoCliente) {
        $this->reservacionMultiple->getDatosCliente()->setDatosCliente($nombre, $apellidoP, $apellidoM, $edad, $tipoCliente);
    }

    public function construirDatosCompra($numTarjeta, $fecha) {
        $this->reservacionMultiple->getDatosCompra()->setDatosCompra($numTarjeta, $fecha);
    }

    public function construirDatosFuncion($asiento, $hora, $dia, $mes, $anio, $pelicula) {
        $this->reservacionMultiple->getDatosFuncion()->setDatosFuncion($asiento, $hora, $dia, $mes, $anio, $pelicula);
    }
    
    public function construirDatosBeneficio($beneficios) {
    if (is_null($beneficios)){
            $this->reservacionMultiple->getDatosBeneficio()->setDatosBeneficio($beneficios);
            
            $reservacion = new ReservacionVIP();
            $reservacion->setDatosCliente($this->reservacionMultiple->getDatosCliente());
            $reservacion->setDatosCompra($this->reservacionMultiple->getDatosCompra());
            $reservacion->setDatosFuncion($this->reservacionMultiple->getDatosFuncion());
            $reservacion->setDatosBeneficio($this->reservacionMultiple->getDatosBeneficio());
            $this->reservacionMultiple->agregarReservacion($reservacion);
        }else{
            $reservacion = new ReservacionSencilla();
            $reservacion->setDatosCliente($this->reservacionMultiple->getDatosCliente());
            $reservacion->setDatosCompra($this->reservacionMultiple->getDatosCompra());
            $reservacion->setDatosFuncion($this->reservacionMultiple->getDatosFuncion());
            
            $this->reservacionMultiple->agregarReservacion($reservacion);
        }
        
    }

    public function obtenerReservacion() {
        return $this->reservacionMultiple;
    }

}
